[SPARK-4666] Improve YarnAllocator's parsing of "memory overhead" param

ryan-williams opened the pull request:
- let it be specified as a fraction of the executor memory
- add/generalize some utilities for parsing "memory strings" that are found around the Spark codebase.

fixes SPARK-4665 and SPARK-4666.

ryan-williams committed:
memoryStringToMb can have default scale specified

Previously it assumed a unitless number represented raw bytes, but I
want to use it for a config variable that previously defaulted to # of
megabytes and not break backwards-compatibility.

srowen commented:
FWIW I think the logic behind the existing parameters is: 7% should be enough for anyone. And if it isn't, they know it, and have a clear idea of just how much they want the overhead to be. Or: if 7% is consistently not enough, it should be higher, like 10%. I'd leave the existing param unless it's obviously problematic.

But that's separate from the question here, of whether lots of these strings could be specified in mega or gigabytes. Is that still something that can move forward? the refactoring of all that seems helpful in any event.

sryza commented:
Not a huge deal, but I think the converse is kind of true.  In my experience, the right percentage is totally unclear to everyone in all situations, so pretty much always needs to be set experimentally.  If I settle on a percentage that works for my job (or all my jobs), it would be nice for me not to need to reset this param every time I tweak spark.executor.memory.

srowen commented:
OK maybe there's enough momentum for ... two flags? one of which overrides the other? But maybe we should also bump up the default a bit too.

lianhuiwang commented:
i think memoryOverheadFraction is good to me.

tgravescs commented:
I'm still not on board with having 2 different configs.  You then have to explain to the user what they are,  which one takes presendence, etc. I can see cases the % is useful but then again you can argue the other way too.  If my job needs 4G overhead, I may want to increase the executor memory without changing the overhead amount.  In that case then I have to go mess with % to get it back what I want.  Either way I think the user has to experiment to get the best number.  I haven't seen any reason to bump up the default at this point  but if others have more real world data lets consider it.

How many people are complaining or having real issues with this?  

The main reason I'm against adding this is that I consider it an api change and we now have to support 2 configs until we can get rid of the deprecated one.  Its just more dev overhead, testing, and potential user confusion.   If its enough of a benefit to the user to change it then I would be ok with it but would rather deprecate the existing one in favor of the %.  To me the raw number is more flexible then the % but I see the % as being easier in situations.

lianhuiwang commented:
before i think that both of two configs can be existed.  from @tgravescs i think overhead is more necessary than OverheadFraction , because at some time it has very large memory,we just use overhead to increase executor memory. now memoryOverheadFraction  is very large and inappropriate.

ryan-williams commented:
Thanks for resurrecting this, @srowen. I think there are 3 potentially separable changes here, in rough order of increasing controversial-ness:

1. Make `spark.yarn.executor.memoryOverhead` take a "memory string" (e.g. `1g` or `100m`)
    * defaulting to `m` as the unit for backwards-compatibility.
1. General refactoring of "memory string" handling in conf parameters in some other parts of the codebase.
1. Addition of `spark.yarn.executor.memoryOverheadFraction`

If people prefer, I can make a separate PR with 1∪2 or individual PRs for each of 1 and 2.

Re: 3, I am sympathetic to @tgravescs's arguments, but worry that we are conflating [wishing Spark's configuration surface area was smaller] with [forcing users to build custom Sparks to get at real, existing niches of said configuration surface area (in this case to change a hard-coded `.07` that reasonable people can disagree on the ideal value of)]. Additionally, the arguments about possible confusion around interplay between `memoryOverhead` and `memoryOverheadFraction` are applicable to the status quo, imho.

However, I don't mind much one way or another at this point, so I'm fine dropping 3 if that's what consensus here prefers.

ryan-williams commented:
(until I hear otherwise I am pulling/rebasing 1∪2, roughly the SPARK-4666 bits, into their own PR)

srowen commented:
Without clear consensus at this point, and given 3 is really a question of how you like to override a default in the same way you override others, I'd do 1+2 for the moment.

ryan-williams committed:
fix deprecation warning

ryan-williams commented on core/src/main/scala/org/apache/spark/scheduler/local/LocalBackend.scala:
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.scheduler.local
 import java.nio.ByteBuffer
 
 import scala.concurrent.duration._
+import scala.language.postfixOps
```
Saw this while compiling:

> [WARNING] /Users/ryan/c/spark/streaming/src/main/scala/org/apache/spark/streaming/util/WriteAheadLogManager.scala:156: postfix operator second should be enabled
by making the implicit value scala.language.postfixOps visible.
This can be achieved by adding the import clause 'import scala.language.postfixOps'
or by setting the compiler option -language:postfixOps.
See the Scala docs for value scala.language.postfixOps for a discussion
why the feature should be explicitly enabled.

srowen replied:
Yes that's the right thing, if you think that the use of postfix ops is justified.

ryan-williams committed:
rename: executorMemory -> executorMemoryMB

ryan-williams committed:
rename: amMemory -> amMemoryMB

ryan-williams committed:
rename: driverMemory -> driverMemoryMB

ryan-williams committed:
rename: DEFAULT_MEMORY -> DEFAULT_MEMORY_MB

ryan-williams committed:
rename: amMemoryOverhead -> amMemoryOverheadMB

ryan-williams committed:
rename: executorMemoryOverhead -> executorMemoryOverheadMB

ryan-williams committed:
rename: amMemoryOverhead -> amMemoryOverheadMB

ryan-williams committed:
rename: executorMemoryOverhead -> executorMemoryOverheadMB

ryan-williams committed:
rename: MEMORY_OVERHEAD_MIN -> MEMORY_OVERHEAD_MIN_MB

ryan-williams committed:
rename: executorMemory -> executorMemoryMB

ryan-williams committed:
rename: executorMemory -> executorMemoryMB

ryan-williams committed:
rename: memoryOverhead -> memoryOverheadMB

ryan-williams committed:
rename: memory -> memoryMB

ryan-williams committed:
rename: mem -> memMB

ryan-williams committed:
move getMaxResultSize from Utils to SparkConf

ryan-williams committed:
privatize amMemoryOverheadConf

ryan-williams committed:
refactor memory-size order-of-magnitude constants

ryan-williams committed:
add memory-string-parsing helpers to Utils

ryan-williams committed:
add `getMemory`, `getMB` helpers to SparkConf

ryan-williams committed:
use SparkConf.getMB helper in Yarn memory parsing

ryan-williams committed:
update docs about YARN memory overhead parameters

ryan-williams commented:
OK, this is ready to go again. Changes:
* removed SPARK-4665 from the title and closed that JIRA
* in the process of rebasing this change I found myself wanting "MB" suffixes on various `Int` variables that represent numbers of megabytes; I've included several commits here that perform such renames, but they're separate and easy to remove from this PR if that's controversial
* I made `Utils.getMaxResultSize(SparkConf)` a method of `SparkConf` instead.
* I cleaned up the semantics around which `Utils` and `SparkConf` methods assume `Int`s to represent numbers of megabytes, vs. ones that are generic across memory-size orders-of-magnitude.

Let me know what you think.

srowen commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -203,6 +203,28 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging {
     getOption(key).map(_.toBoolean).getOrElse(defaultValue)
   }
 
+  // Limit of bytes for total size of results (default is 1GB)
```
Hm, why is this property special-cased here? the methods in this class are generally generic.

ryan-williams replied:
I weighed two options:

1. have a method in `Utils` that takes a `SparkConf` as a parameter and thinly wraps a method call on said `SparkConf`
1. make the aforementioned wrapper a method on `SparkConf` that delegates to another method on `SparkConf`

..and felt like the latter was better/cleaner. My feeling was that a kitchen-sink / generically-named `Utils` class that wraps methods for `SparkConf` (and possibly other classes?) to maintain an illusion of simplicity in the `SparkConf` API is not helping code clarity.

Of course, this is subjective and I'm open to putting it back in `Utils`, lmk.

ryan-williams replied:
NB: that was an answer to why this property is special-cased here, as opposed to over in `Utils`. You may be more interested in the question of why it's special-cased at all, but that seems reasonable to me given the couple magic strings and its greater-than-1 call-sites (namely: 2).

srowen replied:
I would have picked `Utils` I suppose. Or is there somewhere less generic to put this that covers the 2 call sites? Other opinions?

ryan-williams replied:
Option 3. could be: Put such methods in a `SparkConfUtils` object that would be less prone to kitchen-sink-iness.

I'm impartial, I'll let you make the call between these three.

srowen replied:
I'd like another set of eyes on the change from @pwendell or @JoshRosen  . The reason I hesitate before merging is the large number of small changes and merge conflict potential. Although I do definitely think the variable names are improved by the suffix.

For this particular issue, maybe expose just `getMemory` from this class, and inline the two simple calls to it that currently use `getMaxResultSize`? writing `getMemory("spark.driver.maxResultSize", "1g", outputScale = 'b')` in two places isn't bad versus constructing a new home for it.

ryan-williams replied:
hm, I'd vote we put it back in `Utils` over un-factoring those two calls that are doing the same thing as one another

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -980,33 +980,77 @@ private[spark] object Utils extends Logging {
+        (lower.substring(0, lower.length - 1).toLong, lastChar)
+      }
+
+    (for {
```
Why is a for construction used here? just to handle the invalid in / out scale param? My personal taste would be to just check that the Option[Long] exists directly and do away with it. I can't tell how much that's just me versus how the kids talk in Scala these days. A looping construct surprised me as there is no loop.

ryan-williams replied:
> Why is a for construction used here? just to handle the invalid in / out scale param?

`for` comprehensions are commonly used when `map`ing over 2 or more monads to avoid arguably-ugly `.flatMap`-chaining; the syntax just removes some boilerplate, e.g.:

```
    scaleCharToFactor.get(inputScaleChar).flatMap(inputScale => 
      scaleCharToFactor.get(outputScaleChar).map(outputScale =>
        inputScale * num / outputScale
      )
    ).getOrElse(
        // throw
      )
```
vs.

```
    (for {
      inputScale <- scaleCharToFactor.get(inputScaleChar)
      outputScale <- scaleCharToFactor.get(outputScaleChar)
    } yield {
      inputScale * num / outputScale
    }).getOrElse(
        // throw
      )
```

(I collapsed the `scale` wrapper line in the latter for apples-to-apples brevity, and can do that in the PR as well).

So, it's not a "looping construct" so much as a "mapping" one, commonly used on `Option`s, `List`s, and [even things like twitter `Future`s](https://twitter.github.io/scala_school/finagle.html) (search for "for {"). 

However, it does get better as the number of chained `map`s increases, e.g. especially when there are 3 or more, so I'm not that tied to it here.

Of course, handling such situations using a `match` is also possible:

```
    (scaleCharToFactor.get(inputScaleChar), scaleCharToFactor.get(outputScaleChar)) match {
      case (Some(inputScale), Some(outputScale)) =>
        inputScale * num / outputScale
      case _ =>
        // throw
    }
```

I prefer all of these to, say, the following straw-man that doesn't take advantage of any of the nice things that come from using `Option`s:

```
    if (scaleCharToFactor.get(inputScaleChar).isDefined && 
        scaleCharToFactor.get(outputScaleChar).isDefined)
      scaleCharToFactor.get(inputScaleChar).get * num / scaleCharToFactor.get(outputScaleChar).get
    else
      // throw
```

but I'll defer to you on the approach you like best.

srowen replied:
Yeah I get what it does. I was comparing in my mind to...

```
val inputScale = scaleCharToFactor.get(inputScaleChar)
val outputScale = scaleCharToFactor.get(outputScaleChar)
require(inputScale.isDefined)
require(outputScale.isDefined)
inputScale.get * num / outputScale.get
```

Here's another instance where I wouldn't mind hearing another opinion as it's a good more general style question.

srowen commented on docs/running-on-yarn.md:
```diff
@@ -113,9 +113,9 @@ Most of the configs are the same for Spark on YARN as for other deployment modes
 <tr>
  <td><code>spark.yarn.executor.memoryOverhead</code></td>
-  <td>executorMemory * 0.07, with minimum of 384 </td>
+  <td>executorMemory * 0.07, with minimum of 384 megabytes </td>
```
Indeed, and should the other properties here get the same treatment, in docs and code?

ryan-williams replied:
good catch, updated the other similar strings in this file

srowen commented:
I personally definitely like the clarity this is adding in the code. Looks like it needs another rebase and test to see where we stand with Jenkins and Mima.

ryan-williams committed:
code review feedback

ryan-williams commented:
OK I think it is back to you @srowen

ryan-williams commented:
(thanks for the review!)

hammer commented:
@srowen this one ready to go in?

srowen commented:
@hammer @ryan-williams I'd prefer to hear from @JoshRosen or @pwendell before merging. The two key questions are

- Gut feeling about the potential for merge conflict -- OK?
- An opinion on https://github.com/apache/spark/pull/3525/files#r24632812 and maybe the `for` comprehension question though that's really not that important either way

JoshRosen commented:
I think that this PR's motivation may have been addressed by #5574.  @ilganeli @srowen @ryan-williams , can you confirm whether there's still an issue here and resolve this PR / JIRAs as appropriate?

srowen commented:
I think this would have to be reworked in light of SPARK-5932. I don't think that resolved this particular suggestion. Proceed that way or close this PR?

ilganeli commented:
@srowen @JoshRosen I think this should be refactored to use the updates from #5574 but I don't think #5574 resolves this on its own because of the need to handle the min/max allocation - my 2c.

andrewor14 commented:
@ryan-williams would you mind opening a new PR to reflect the changes in #5574?

ryan-williams commented:
I've lost state on this, closing, thanks all
